### PR TITLE
binutils: unbreak w/musl via upstream commits

### PR DESCRIPTION
**Pull request by dtzWill**
###### Motivation for this change

https://github.com/NixOS/nixpkgs/commit/53e1db960a0663313e950f4ad45c98a82ad7d2c3#commitcomment-32020416

Try building 'busybox-sandbox-shell' without these fixes,
observe crash :(.

Try building 'busybox-sandbox-shell' **with** these fixes,
observe no crash! \o/

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dtzWill** [comment]
cc @vcunat @NeQuissimus (mentioning the commit didn't seem to 'ping' that discussion? dunno)

**Comment by vcunat** [comment]
Yes, GitHub doesn't process mentions to commits.

**Comment by NeQuissimus** [comment]
@GrahamcOfBorg build busybox-sandbox-shell

**Comment by dtzWill** [comment]
Yes, I was trying to submit what I thought would get accepted/fixed
first... to unbreak my things xD.

(and with reproducible and simple example of breakage/fixing)

One would hope with that many changes/fixes they'll cut a new release
soon? Wishful thinking, of course, gah.

If it wasn't so rebuild-costly I'd in fact suggest we follow it closely,
after vetting them and such.

Hmm, anyone know /offhand/ how other distributions handle this? O:)

**Comment by NeQuissimus** [comment]
Ubuntu bundles a ton of patches: https://packages.ubuntu.com/cosmic/all/binutils-source/filelist

**Comment by vcunat** [comment]
I believe we can easily afford a stdenv rebuild every month.  Human time is more of a problem for me here &ndash; just deciding whether patches are mutually independent (and I don't mean failures of applying them) and which of them matter for us.

**Comment by dtzWill** [comment]
Okay to merge as-is to get this fixed (things are broken currently!), and we can "upgrade" binutils further as a separate PR/issue?  I'd say RFC but that seems a bit heavy for this sort of discussion :).

File: pkgs/development/tools/misc/binutils/default.nix [hunk 1]
```diff
@@ -54,6 +54,12 @@ stdenv.mkDerivation rec {
 
     # https://sourceware.org/bugzilla/show_bug.cgi?id=22868
     ./gold-symbol-visibility.patch
+
+    # https://sourceware.org/bugzilla/show_bug.cgi?id=23428
+    # un-break features so linking against musl doesn't produce crash-only binaries
+    ./0001-x86-Add-a-GNU_PROPERTY_X86_ISA_1_USED-note-if-needed.patch
+    ./0001-x86-Properly-merge-GNU_PROPERTY_X86_ISA_1_USED.patch
+    ./0001-x86-Properly-add-X86_ISA_1_NEEDED-property.patch
```
Comment (Mic92) [inline pkgs/development/tools/misc/binutils/default.nix]: Is it possible to use fetchurl for any of these or did you patch them locally?

**Comment by dtzWill** [comment]
Wasn't sure fetchpatch would work re:bootstrap, can probably grab via fetchurlBoot but would need some fetchpatch (filterdiff) alternative to drop the changelog portions of the commits (I kept the commits separate in this PR to help make that clear, esp looking back later on).

Anyway maybe we can, but not sure how reasonable it'd be.  Maybe there's a nice way to do it, in which case I agree that's be preferable.

**Comment by vcunat** [comment]
I think I tried hard to find a way when patching glibc with upstream commits, some time ago, and I failed.  I ended up putting large gzipped diffs into nixpkgs.

**Pull request merged**